### ISSUE TITLE
sql: unexpected RELEASE crashes node

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#24433]
Doing `begin; release savepoint cockroach_restart;` results in:

```
panic while executing 1 statements: RELEASE SAVEPOINT cockroach_restart; caused by event TxnReleased{} inappropriate in current state Open{ImplicitTxn:false, RetryIntent:false}

goroutine 599 [running]:
runtime/debug.Stack(0x6134560, 0xc4213b7b40, 0x3)
	/Users/andrei/go/src/runtime/debug/stack.go:24 +0xa7
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x6134560, 0xc4213b7b40, 0xc4206a7980, 0x5d279a0, 0xc420f5d9e0, 0x1)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:195 +0xa5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc4207d0000, 0x6134560, 0xc4213b7b40, 0x5c7fea0, 0xc420424a60)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:521 +0x2d5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc4207d0000, 0x6134560, 0xc4213b7b40)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0x61
panic(0x5c7fea0, 0xc420424a60)
	/Users/andrei/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).txnStateTransitionsApplyWrapper(0xc4207d0000, 0x6115fa0, 0x7335888, 0x0, 0x0, 0x921ca78, 0xc420f3ca00, 0xd, 0x0, 0x0, ...)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1649 +0x6dc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc4207d0000, 0x6134560, 0xc4213b7b40, 0x0, 0x0)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1029 +0xb1e
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc420276380, 0x6134560, 0xc4213b7b40, 0xc420ae004a, 0x4, 0xc420ae0054, 0x4, 0x0, 0x0, 0x611d860, ...)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491 +0xf80
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc420276380, 0x6134560, 0xc4213b7b40, 0xc4212cfc00, 0x5400, 0x15000, 0xc4209ba190, 0xc4203ca0e0, 0xc420309fb0, 0xc4203ca0d0)
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0xfc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/Users/andrei/work2/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf1
```

**Comment by tim-o** [comment]
Hey @andreimatei - was this fixed in 2.0? There's a new example that just came through sentry, v2.0.2:

https://sentry.io/cockroach-labs/cockroachdb/issues/579110012/?referrer=slack

**Comment by andreimatei** [comment]
#25251 was cherry-picked into `v2.0.2`. I can teach you how to tell these things if you want.

This report is a different bug. Filed #26687.
